clang/lib/AST/ExprConstant.cpp

```diff
@@ -1703,7 +1703,7 @@ namespace {
     bool checkNullPointerDiagnosingWith(const GenDiagType &GenDiag) {
       if (Designator.Invalid)
         return false;
-      if (IsNullPtr) {
+      if (getLValueBase().isNull()) {
```

Comment:
Why was the old check not good enough here? In the added testcase, `nb` is a null pointer.

Reply:
Oh, I see. We're adjusting the offset *before* we call `addDecl`. I think this would be better fixed by reversing the order in which we do those steps -- first `addDecl` and then adjust the offset. A null base isn't necessarily a null pointer, so checking for a null base here seems wrong (but only in the weird case of a zero integer cast to a non-null pointer value, which we should only accept in constant-folding mode, not when properly evaluating).

Reply:
Looks like the check here was `!Base` until 402804b6d67c880c97e575342189025fe28c0f81